control/state.py

```diff
@@ -7,109 +7,179 @@
+
+        # Delete all keys related to subsystem
+        state = self.get_state()
+        for key in state.keys():
+            if (key.startswith(self.NAMESPACE_PREFIX + subsystem_nqn) or
+                    key.startswith(self.HOST_PREFIX + subsystem_nqn) or
+                    key.startswith(self.LISTENER_PREFIX + subsystem_nqn)):
+                self._remove_key(key)
 
-    @abstractmethod
     def add_host(self, subsystem_nqn: str, host_nqn: str, val: str):
-        pass
+        """Adds a host to the state data store."""
+        key = "{}{}_{}".format(self.HOST_PREFIX, subsystem_nqn, host_nqn)
+        self._add_key(key, val)
 
-    @abstractmethod
     def remove_host(self, subsystem_nqn: str, host_nqn: str):
-        pass
+        """Removes a host from the state data store."""
+        key = "{}{}_{}".format(self.HOST_PREFIX, subsystem_nqn, host_nqn)
+        self._remove_key(key)
 
-    @abstractmethod
-    def add_listener(self, subsystem_nqn: str, traddr: str, trsvcid: str,
-                     val: str):
-        pass
+    def add_listener(self, subsystem_nqn: str, gateway: str, trtype: str,
+                     traddr: str, trsvcid: str, val: str):
+        """Adds a listener to the state data store."""
+        key = "{}{}_{}_{}_{}_{}".format(self.LISTENER_PREFIX, gateway,
+                                        subsystem_nqn, trtype, traddr, trsvcid)
+        self._add_key(key, val)
 
-    @abstractmethod
-    def remove_listener(self, subsystem_nqn: str, traddr: str, trsvcid: str):
-        pass
+    def remove_listener(self, subsystem_nqn: str, gateway: str, trtype: str,
+                        traddr: str, trsvcid: str):
+        """Removes a listener from the state data store."""
+        key = "{}{}_{}_{}_{}_{}".format(self.LISTENER_PREFIX, gateway,
+                                        subsystem_nqn, trtype, traddr, trsvcid)
+        self._remove_key(key)
 
     @abstractmethod
     def delete_state(self):
+        """Deletes state data store."""
         pass
 
-    @abstractmethod
-    def restore(self, callbacks):
-        pass
+
+class LocalGatewayState(GatewayState):
+    """Persists gateway NVMeoF target state to a local object.
+
+    Records persistent NVMeoF target state data in key/value format within a 
+    local object.
+
+    Instance attributes:
+        state: Local gateway NVMeoF target state
+    """
+
+    def __init__(self):
+        self.state = {}
+
+    def get_state(self) -> Dict[str, str]:
+        """Returns state object."""
+        return self.state
+
+    def _add_key(self, key: str, val: str):
+        """Adds key and value to the local state object."""
+        self.state[key] = val.encode('utf-8')
+
+    def _remove_key(self, key: str):
+        """Removes key from the local state object."""
+        self.state.pop(key)
+
+    def delete_state(self):
+        """Deletes contents of local state object."""
+        self.state.clear()
+
+    def update(self, state_update):
```

Comment:
Let's avoid confusion with `ConsistentGatewayState.update` and make the signature reflect how this method is actually used:
```suggestion
    def reset(self, omap_state):
```

Reply:
👍